### PR TITLE
Feature/typescript functions improvements

### PR DESCRIPTION
**Pull request by leon**
### Description
Updates and improves the typescript templates
The default for typescript was outdated, and I thought it would be best to update to node 8 and typescript 3.

Let me know if you have any questions :)

### Commits

* add src to the functions.ignore settings on init
* include @firebase/app-types
* update typescript to version 3.1.4
* ignore node_modules/ and compiled lib/ directory
* set default of new functions project to node 8
	 
### Scenarios Tested

Tests run
Tested to link firebase-tools and init a firebase functions project and then compile and lint that.

### CODE REVIEW

File: templates/init/functions/typescript/package.lint.json [hunk 1]
```diff
@@ -9,14 +11,16 @@
...
   },
-  "main": "lib/index.js",
   "dependencies": {
     "firebase-admin": "~6.0.0",
     "firebase-functions": "^2.1.0"
   },
   "devDependencies": {
-    "tslint": "~5.8.0",
-    "typescript": "~2.8.3"
+    "@firebase/app-types": "^0.3.2",
```
Comment (bkendall) [inline templates/init/functions/typescript/package.lint.json]: Why is this necessary to include?

**Comment by bkendall** [comment]
Thanks for taking the time to put this together.

The biggest flag for me is that this is now wanting node `8`, though it is [still in](https://cloud.google.com/functions/docs/concepts/nodejs-8-runtime) [beta](https://firebase.google.com/docs/functions/manage-functions#set_nodejs_version). Because of that, the templates we provide should match our compatibility for the lowest node version we support, in this case: `6`.

Along the same vein, we probably don't want to target the build to ES2017 as a lot of those features won't work in node `6`. The target should remain most likely at `ES6`/`ES2015`.

File: templates/init/functions/typescript/package.lint.json (hunk 1 above)
Comment (leon) [reply templates/init/functions/typescript/package.lint.json]: Before I was having troubles with the firebase typings, but I think that might have gotten fixed with the 6.0.0 release.

I'll remove it.

**Comment by leon** [comment]
I understand.
Though I find it hard to believe anyone starting a new firebase project would want to use node 6 since it's dead.
There is even a node 10 LTS now.
https://github.com/nodejs/Release

es2017 includes a lot of good features, and using async/await with node8 makes writing the cloud functions so much easier.

If we are not ready to move to node8 yet, we can wait with merging this until node8 comes out of beta.
Because most of it is to make it easier for people to get a node8 project up and running.

**Comment by bkendall** [comment]
By [your source](https://github.com/nodejs/Release), Node 6 isn't dead. Its EoL is April 2019.

I do agree that some of the newer features are great to work with, but since we still need to support Node 6, we will probably keep that as the default for the time being. We will consider what we do when Node 8 is out of beta (and Node 6 EoL starts looming over us), but I don't see this changing before the new year.

Thank you for the update though - I think keeping this in the back pocket will be good for when Node 8 hopefully comes out of beta for functions!

**Comment by oBusk** [comment]
The changes in this PR is still good, even without the `es2017` change. 
Getting outdated versions of typescript and/or tslint when using firebase tools to setup a project is not helpful: Any plans on removing that change and merging this in?

**Comment by oBusk** [comment]
> I think a way to solve these concerns would be to simply prompt the user what is their desired [runtime](https://cloud.google.com/functions/docs/concepts/exec#runtimes), and then initialize the appropriate templates for that choice.

It's arguable that it would make the setup process more confusing than necessary. Once you've created the starting template, upgrading to node 8 isn't that hard. And once node 8 runtime is not deemed fully released, the template can be upgraded. If you're not working a lot with node, having any clue if you want node 6 or node 8 is hard as a first step when starting a project.

I still think that the typescript templates should be updated, and this PR should be modified or a new one should be created.

**Comment by bkendall** [comment]
I'm going to put this on my stack for this week and will try to get back here in a day or two. Thanks for y'all's patience...

**Comment by bkendall** [comment]
So I incorporated a number of these updates into #1093, which has been submitted and released. We're not ready to update the default to Node 8 _quite yet_, which is why we couldn't take this PR whole. Thank you for the work identifying the things that needed to get updated! If you want to give the CLI release `v6.3.0` a try, you should see some updated initial states when you create a Functions project!
